commons/src/layer/cache_buddy.rs

```diff
@@ -0,0 +1,482 @@
+//! in a user readable format.
+//!
+//! ## Invalid metadata migration ([`TryMigrate`])
+//!
+//! If previously serialized metadata cannot be deserialized into the current struct then usually the
+//! only thing a buildpack can do is discard the cache. However, that may involve needing to re-do an
+//! expensive operation such as re-compiling native libraries. Buildpack authors should feel free to
+//! refactor and update their metadata structs without fear of busting the cache. Users should not
+//! have to suffer slower builds due to internal only buildpack changes.
+//!
+//! The `TryMigrate` trait from `magic_migrate` allows buildpack authors to define how to migrate an
+//! older struct into a newer one. If the migration fails, the cache is cleared and the reason is returned.
+//! If the migration succeeds, then the regular logic in `CacheDiff` is applied.
+//!
+//! ## Read your write, or (read) why you can't ([`Meta`])
+//!
+//! If non-cache data is stored in the Metadata, then your buildpack may want to read that data back.
+//! When the cache is not cleared then the old metadata is returned. This allows you to read your write.
+//!
+//! A buildpack cache should never be cleared without explaining why to a user via printing to the
+//! build output. If the cache is cleared for any reason, then a user readable message is returned. This message should
+//! be printed to the buildpack user so they can understand what caused the cache to clear.
+//!
+#![doc = include_str!("./fixtures/cache_buddy_example.md")]
+
+use crate::display::SentenceList;
+use cache_diff::CacheDiff;
+use libcnb::build::BuildContext;
+use libcnb::data::layer::LayerName;
+use libcnb::layer::{CachedLayerDefinition, InvalidMetadataAction, LayerRef, RestoredLayerAction};
+use magic_migrate::TryMigrate;
+use serde::ser::Serialize;
+use std::fmt::Debug;
+
+/// Handle caching behavior for a layer
+///
+/// Guarantees that new metadata is always written (prevents accidentally reading one struct type and
+/// writing a different one). It also provides a standard interface to define caching behavior via
+/// the [`CacheDiff`] and [`TryMigrate`] traits:
+///
+/// - The [`TryMigrate`] trait is for handling invalid metadata:
+///   When old metadata from cache is invalid, we try to load it into a known older version and then migrate it
+///   to the latest via `TryMigrate`. If that fails, the layer is deleted and the error is returned. If it
+///   succeeds, then the logic in `CacheDiff` below is applied.
+///
+/// The [`CacheDiff`] trait defines cache invalidation behavior when metadata is valid:
+///   When a `CacheDiff::diff` is empty, the layer is kept and the old data is returned. Otherwise,
+///   the layer is deleted and the changes are returned.
+///
+#[doc = include_str!("./fixtures/cache_buddy_example.md")]
+#[derive(Default, Debug, Clone, Eq, PartialEq)]
+pub struct CacheBuddy {
```

Comment:
I'm not sure `CacheBuddy` is a great name for this struct -- it doesn't seem immediately obvious that the build and launch fields here refer to the layer definition (particularly since the semantics of those fields also differ)